module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -531,63 +543,81 @@ namespace app
+        // THIS IS IN TEST!
+        const auto &[name, data] = msg->getSwitchData();
+        pushWindow(name, data);
+        getCurrentWindow()->handleSwitchData(switchData.get());
+
+        if (handleUpdateTextRefresh(switchData.get())) {
+            return sys::msgHandled();
+        }
+        getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
 
-            auto ret = dynamic_cast<gui::SwitchSpecialChar *>(switchData.get());
-            if (ret != nullptr && ret->type == gui::SwitchSpecialChar::Type::Response) {
-                auto text = dynamic_cast<gui::Text *>(getCurrentWindow()->getFocusItem());
-                if (text != nullptr) {
-                    text->addText(ret->getDescription());
-                    refreshWindow(gui::RefreshModes::GUI_REFRESH_FAST);
-                    return sys::msgHandled();
-                }
-            }
-            getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
+        /// This is kind of gentle hack on our stateless Application Window management switching
+        /// if we are not requesting for popup and we cant handle next popup - just refresh the window
+        /// else -> popup will refresh it for us
+        if (not tryShowPopup()) {
             refreshWindow(gui::RefreshModes::GUI_REFRESH_DEEP);
         }
-        else {
-            LOG_ERROR("No such window: %s", msg->getWindowName().c_str());
-        }
         return sys::msgHandled();
     }
 
+    bool ApplicationCommon::handleUpdateTextRefresh(gui::SwitchData *data)
+    {
+        auto ret = dynamic_cast<gui::SwitchSpecialChar *>(data);
+        if (ret != nullptr && ret->type == gui::SwitchSpecialChar::Type::Response) {
+            auto text = dynamic_cast<gui::Text *>(getCurrentWindow()->getFocusItem());
+            if (text != nullptr) {
+                text->addText(ret->getDescription());
+                refreshWindow(gui::RefreshModes::GUI_REFRESH_FAST);
+                return true;
+            }
+        }
+        return false;
+    }
+
     sys::MessagePointer ApplicationCommon::handleUpdateWindow(sys::Message *msgl)
     {
         auto msg = static_cast<AppUpdateWindowMessage *>(msgl);
-
-        if (windowsFactory.isRegistered(msg->getWindowName()) && isCurrentWindow(msg->getWindowName())) {
+        auto haveBuilder = windowsFactory.isRegistered(msg->getWindowName());
+        auto window      = isCurrentWindow(msg->getWindowName());
+        if (haveBuilder && window) {
             const auto &switchData = msg->getData();
             getCurrentWindow()->handleSwitchData(switchData.get());
             getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
             refreshWindow(msg->getRefreshMode());
         }
         else {
-            LOG_ERROR("No such window: %s", msg->getWindowName().c_str());
+            LOG_ERROR("Wont update window: %s in app: %s haveBuilder: %s is_on_top: %s",
```

Comment:
Please rephrase it if it need to be an useful error  message ;)

Reply:
what's not useful here exactly? :)